### PR TITLE
Improve FileModes check

### PR DESCRIPTION
**Pull request by racodond**
- Fix false negatives on Resource Default Statements
- Fix false negatives on mode set to integer or octal values
- Double-quoted strings should not be allowed
- Deprecate unquoted_file_mode rule and add links to the replacement rule
- Extend FilesMode rule description
- Fix false positives on string containing variables
- Tag FilesModes as BUG
- Add reference to the Puppet Las Puppet Language Style Guide, Quoting section
- Add Noncompliant comments to file_mode.pp to quickly find out where issues are expected to be raised

### CODE REVIEW

**Comment by racodond** [comment]
@iwarapter: I'll let you review it.

Also, could we work by pull requests so that we can review each other's work? Especially for rules because there are usually a ton of corner cases that you don't think about at first. And it leads to a ton of false positives and false negatives.

**Comment by iwarapter** [comment]
@racodond i'll take a look when i get home, sorry about not doing the pull request - old habits :(

**Comment by racodond** [comment]
> i'll take a look when i get home

Cool. Thanks!

> sorry about not doing the pull request - old habits :(

No problem, it was just a reminder :-)

**Comment by racodond** [comment]
@iwarapter Also, as a nice habit, we could start to add 'Noncompliant' comments on test files to quickly find out where issues are expected to be raised. See https://github.com/iwarapter/sonar-puppet/commit/7979191a95b673128060a2588ea0e22e49846f37

**Comment by iwarapter** [comment]
@racodond Yeah that is a good idea.

File: puppet-checks/src/main/java/com/iadams/sonarqube/puppet/checks/FileModeCheck.java [hunk 1]
```diff
@@ -60,22 +63,36 @@ public void init() {
 
   @Override
   public void visitNode(AstNode node) {
-    if(node.getTokenValue().equals("file")) {
+    if ("file".equals(node.getTokenValue())) {
       for (AstNode body : node.getDescendants(PuppetGrammar.RESOURCE_INST)) {
```
Comment (racodond) [inline puppet-checks/src/main/java/com/iadams/sonarqube/puppet/checks/FileModeCheck.java:67]: @iwarapter You should be very very careful while using `getDescendants` especially on nested structures. Would it be possible to replace them by `getChildren` here?

Comment (iwarapter) [inline puppet-checks/src/main/java/com/iadams/sonarqube/puppet/checks/FileModeCheck.java:67]: I'd need to double check if the RESOURCE_INST are first level children but probably, i'll check later.

Comment (racodond) [inline puppet-checks/src/main/java/com/iadams/sonarqube/puppet/checks/FileModeCheck.java:67]: @iwarapter After having a look at the grammar, I now understand why you extensively use `getDescendants` instead of `getChildren`. It's because the generated AST is too verbose. Quite a few `skip()` should  be added to the grammar. See for example my changes on the grammar in the following commit: https://github.com/iwarapter/sonar-puppet/commit/5ee0c4712ac1e9adbe5074281f0e2d699195939c
The way you wrote the grammar is nice because it is easy to read. But the layers you added (`SVALUES`, `SINTVALUES`, `SELECTLHAND` for example) are very annoying while writing checks. That's why they should be "skipped" while building the AST. In theory, all nodes should be skipped at first, and they should be "unskipped" once they are needed to write checks (or metrics). We can go the other way around if you want and skip the layers that are "annoying" while we write new checks.
To summarize, if you need to use `getDescendants`, it's likely that some nodes should be skipped.

Comment (iwarapter) [inline puppet-checks/src/main/java/com/iadams/sonarqube/puppet/checks/FileModeCheck.java:67]: @racodond Yes, i have modelled the grammar almost verbatim after the ruby implementation, i'm happy to add skips where they add value (I wasn't sure sure if there are any best practices around maintaining the AST vs easier checks etc).
